File: src/Research.FSharp/RenderedSignalCarrier.fs
namespace Zeta.Research

open System
open System.Diagnostics
open System.Security.Cryptography
open Zeta.Core

/// A source-owned animation carrier. Only decoded rendered observations enter learners.
[<RequireQualifiedAccess>]
module RenderedSignalCarrier =
    type Renderer = TrainDot | HeldoutBar | Nuisance
    type Diagnostics = { Comparisons: int; StructureChanges: int; PaletteChanges: int; PlacementChanges: int }
    type Fingerprints = { RomSha256: string; FrameSha256: string; TokenSha256: string; Sequences: int; Frames: int }
    type Resource = { ElapsedMilliseconds: float; CpuMilliseconds: float; AllocatedBytes: int64 }
    type Corpus = { Tokens: int[][]; Fingerprints: Fingerprints; Diagnostics: Diagnostics; Generation: Resource; Extraction: Resource }

    let name = function TrainDot -> "train-dot" | HeldoutBar -> "heldout-bar" | Nuisance -> "nuisance"
    let renderer = function "train-dot" -> Ok TrainDot | "heldout-bar" -> Ok HeldoutBar | "nuisance" -> Ok Nuisance | _ -> Error "unknown rendered carrier"
    let sha256 (bytes: byte[]) = SHA256.HashData bytes |> Convert.ToHexString
    let zeroDiagnostics = { Comparisons = 0; StructureChanges = 0; PaletteChanges = 0; PlacementChanges = 0 }
    let addDiagnostics a b =
        { Comparisons = a.Comparisons + b.Comparisons; StructureChanges = a.StructureChanges + b.StructureChanges
          PaletteChanges = a.PaletteChanges + b.PaletteChanges; PlacementChanges = a.PlacementChanges + b.PlacementChanges }
    let measure action =
        use proc = Process.GetCurrentProcess()
        let cpu = proc.TotalProcessorTime
        let allocated = GC.GetAllocatedBytesForCurrentThread()
        let timer = Stopwatch.StartNew()
        let value = action ()
        timer.Stop()
        value, { ElapsedMilliseconds = timer.Elapsed.TotalMilliseconds; CpuMilliseconds = (proc.TotalProcessorTime - cpu).TotalMilliseconds
                 AllocatedBytes = GC.GetAllocatedBytesForCurrentThread() - allocated }

    let sample (stream: ResearchRandom.Stream) length probability start duration =
        if length < 2 || length > 256 || not (Double.IsFinite probability) || probability < 0.0 || probability > 1.0
           || start < 2 || duration < 0 || (duration > 0 && (start >= length || duration > length - start)) then
            Error "source requires 2..256 symbols and a bounded change interval"
        else
            let tokens = Array.zeroCreate length
            for t in 0 .. length - 1 do
                let draw = stream.Next()
                if t < 2 then tokens.[t] <- int (2.0 * draw)
                else
                    let p = if t >= start && t < start + duration then probability else 0.75
                    tokens.[t] <- if draw < p then tokens.[t - 2] else 1 - tokens.[t - 2]
            Ok tokens

    let private layout renderer count =
        let groupBytes = if renderer = Nuisance then 12 else 8
        groupBytes, 0x204 + groupBytes * count

    let compile renderer (tokens: int[]) =
        if isNull tokens || tokens.Length < 1 || tokens.Length > 256 || Array.exists (fun t -> t <> 0 && t <> 1) tokens then
            Error "carrier requires 1..256 binary symbols"
        else
            let groupBytes, sprite = layout renderer tokens.Length
            let size = sprite - 0x200 + (if renderer = Nuisance then 2 else 1)
            if size > Chip8.MemSize - Chip8.ProgramStart then Error "ROM exceeds classic CHIP-8 program memory"
            else
                let bytes = ResizeArray<byte>()
                let emit opcode = bytes.Add(byte (opcode >>> 8)); bytes.Add(byte (opcode &&& 255))
                emit (0x6100 ||| (if renderer = HeldoutBar then 20 else 8))
                emit (0xA000 ||| sprite)
                for t in 0 .. tokens.Length - 1 do
                    let bar = renderer = HeldoutBar || (renderer = Nuisance && t % 2 = 1)
                    if renderer = Nuisance then
                        emit (0x6100 ||| (if bar then 20 else 8))
                        emit (0xA000 ||| (sprite + (if bar then 1 else 0)))
                    emit 0x00E0
                    emit (0x6000 ||| ((if bar then 14 else 16) + 32 * tokens.[t]))
                    emit 0xD011
                    let next = 0x204 + groupBytes * (t + 1)
                    emit (0x1000 ||| (if t = tokens.Length - 1 then next - 2 else next))
                bytes.Add(if renderer = HeldoutBar then 0xE0uy else 0x80uy)
                if renderer = Nuisance then bytes.Add 0xE0uy
                Ok(bytes.ToArray())

    /// Validate the complete tiny carrier before an opcode reaches the mutable emulator.
    let validateRom renderer count (rom: byte[]) =
        if count < 1 || count > 256 || isNull rom then Error "invalid ROM or frame count"
        else
            let groupBytes, sprite = layout renderer count
            let size = sprite - 0x200 + (if renderer = Nuisance then 2 else 1)
            if rom.Length <> size || rom.Length > Chip8.MemSize - Chip8.ProgramStart then Error "ROM shape or budget mismatch"
            else
                let mutable valid = true
                let tokens = Array.init count (fun t ->
                    let bar = renderer = HeldoutBar || (renderer = Nuisance && t % 2 = 1)
                    let index = 4 + t * groupBytes + (if renderer = Nuisance then 4 else 0) + 3
                    let coordinate = int rom.[index] - (if bar then 14 else 16)
                    if coordinate <> 0 && coordinate <> 32 then valid <- false
                    coordinate / 32)
                if not valid then Error "noncanonical beacon coordinate"
                else
                    compile renderer tokens |> Result.bind (fun expected ->
                        if rom.AsSpan().SequenceEqual(expected.AsSpan()) then Ok() else Error "noncanonical carrier opcode or sprite")

    let decode (frame: GameEnvironment.Frame) =
        if frame.W <> 64 || frame.H <> 32 || frame.Palette <> 2 then Error "beacon requires a 64x32 binary-palette frame"
        else
            FrameSignals.observe frame |> Result.mapError (sprintf "%A") |> Result.bind (fun signals ->
                match signals.Components with
                | [ region ] when region.Origin.X / 32 = (region.Origin.X + region.Shape.Width - 1) / 32 ->
                    FrameMotion.observe FrameMotion.empty frame |> Result.mapError (sprintf "%A") |> Result.bind (fun motion ->
                        match motion.Current with Some point -> Ok(point.X / 32) | None -> Error "missing beacon centroid")
                | _ -> Error "beacon must be one component entirely within one horizontal half")

    /// The callback sees copied rendered cells only. Every opcode and resulting PC are checked.
    let renderRom renderer count (rom: byte[]) (onFrame: GameEnvironment.Frame -> unit) =
        validateRom renderer count rom |> Result.bind (fun () ->
            let machine = Chip8.create 1UL
            Chip8.loadRom rom machine
            let groupBytes, _ = layout renderer count
            let tokens = Array.zeroCreate count
            let mutable failure = None
            let mutable previous = None
            let mutable diagnostics = zeroDiagnostics
            let step () =
                if failure.IsNone then
                    let pc = int machine.PC
                    if pc < 0x200 || pc + 1 >= 0x200 + rom.Length then failure <- Some "program counter outside admitted ROM"
                    else
                        let opcode = (int machine.Mem.[pc] <<< 8) ||| int machine.Mem.[pc + 1]
                        let expected = if opcode &&& 0xF000 = 0x1000 then opcode &&& 0xFFF else pc + 2
                        Chip8.step machine
                        if int machine.PC <> expected then failure <- Some "emulator program counter disagrees with admitted opcode"
            step (); step ()
            for t in 0 .. count - 1 do
                for _ in 1 .. groupBytes / 2 do step ()
                if failure.IsNone then
                    let invert = renderer = Nuisance && t % 2 = 1
                    let cells = machine.Display |> Array.map (fun lit -> if lit <> invert then 1uy else 0uy)
                    let frame: GameEnvironment.Frame = { W = 64; H = 32; Palette = 2; Cells = cells }
                    match decode frame with
                    | Error reason -> failure <- Some reason
                    | Ok token ->
                        tokens.[t] <- token
                        match previous with
                        | None -> ()
                        | Some prior ->
                            match FrameSignals.compare prior frame with
                            | Error reason -> failure <- Some(sprintf "%A" reason)
                            | Ok delta -> diagnostics <- addDiagnostics diagnostics
                                                            { Comparisons = 1; StructureChanges = if delta.StructureChanged then 1 else 0
                                                              PaletteChanges = if delta.PaletteChanged then 1 else 0
                                                              PlacementChanges = if delta.PlacementChanged then 1 else 0 }
                        previous <- Some frame
                        onFrame frame
            match failure with Some reason -> Error reason | None -> Ok(tokens, diagnostics))

    /// Corpus truth is used only for extraction conformance, never as fallback input.
    let corpus renderer count length seed domain probability start duration =
        if count < 1 || count > 65536 then Error "corpus count must be in 1..65536"
        else
            let stream = ResearchRandom.Stream(ResearchRandom.domain seed domain)
            let (generated, generation) = measure (fun () ->
                let rows = ResizeArray<int[] * byte[]>()
                let mutable failure = None
                for _ in 1 .. count do
                    if failure.IsNone then
                        match sample stream length probability start duration |> Result.bind (fun tokens -> compile renderer tokens |> Result.map (fun rom -> tokens, rom)) with
                        | Ok row -> rows.Add row
                        | Error reason -> failure <- Some reason
                match failure with Some reason -> Error reason | None -> Ok(rows.ToArray()))
            generated |> Result.bind (fun generatedRows ->
                use romHash = IncrementalHash.CreateHash HashAlgorithmName.SHA256
                use frameHash = IncrementalHash.CreateHash HashAlgorithmName.SHA256
                use tokenHash = IncrementalHash.CreateHash HashAlgorithmName.SHA256
                let (extracted, extraction) = measure (fun () ->
                    let rows = ResizeArray<int[]>()
                    let mutable failure = None
                    let mutable diagnostics = zeroDiagnostics
                    for truth, rom in generatedRows do
                        if failure.IsNone then
                            romHash.AppendData rom
                            match renderRom renderer length rom (fun frame -> frameHash.AppendData frame.Cells) with
                            | Error reason -> failure <- Some reason
                            | Ok(tokens, delta) ->
                                if tokens <> truth then failure <- Some "rendered extraction disagrees with source truth"
                                else
                                    tokenHash.AppendData(Array.map byte tokens)
                                    rows.Add tokens
                                    diagnostics <- addDiagnostics diagnostics delta
                    match failure with Some reason -> Error reason | None -> Ok(rows.ToArray(), diagnostics))
                extracted |> Result.map (fun (rows, diagnostics) ->
                    { Tokens = rows; Diagnostics = diagnostics; Generation = generation; Extraction = extraction
                      Fingerprints = { RomSha256 = romHash.GetHashAndReset() |> Convert.ToHexString; FrameSha256 = frameHash.GetHashAndReset() |> Convert.ToHexString
                                       TokenSha256 = tokenHash.GetHashAndReset() |> Convert.ToHexString; Sequences = count; Frames = count * length } }))

File: src/Interp.Python/zeta_interp/rendered_signal_carrier.py
"""Independent restricted CHIP-8 beacon interpreter and rendered observations.

This is a reference for the registered opcode subset, not a general emulator.
Generated source symbols never replace observations from executed display cells.
"""

from __future__ import annotations

import hashlib
from collections.abc import Callable

import numpy as np

from zeta_interp.mess3_replay import Stream, domain

RENDERERS = ("train-dot", "heldout-bar", "nuisance")


def sample(stream, length, probability=0.75, start=128, duration=0):
    if (
        not 2 <= length <= 256
        or not np.isfinite(probability)
        or not 0 <= probability <= 1
        or start < 2
        or duration < 0
        or (duration and (start >= length or duration > length - start))
    ):
        raise ValueError("invalid source bounds")
    tokens = np.zeros(length, dtype=np.uint8)
    for t in range(length):
        draw = stream.next()
        if t < 2:
            tokens[t] = int(2 * draw)
        else:
            p = probability if start <= t < start + duration else 0.75
            tokens[t] = tokens[t - 2] if draw < p else 1 - tokens[t - 2]
    return tokens


def compile_rom(renderer, tokens):
    values = np.asarray(tokens)
    if (
        renderer not in RENDERERS
        or values.ndim != 1
        or not 1 <= len(values) <= 256
        or not np.isin(values, [0, 1]).all()
    ):
        raise ValueError("invalid carrier or binary symbols")
    group_bytes = 12 if renderer == "nuisance" else 8
    sprite = 0x204 + group_bytes * len(values)
    opcodes = [0x6100 | (20 if renderer == "heldout-bar" else 8), 0xA000 | sprite]
    for t, token in enumerate(values):
        bar = renderer == "heldout-bar" or (renderer == "nuisance" and t % 2 == 1)
        if renderer == "nuisance":
            opcodes.extend([0x6100 | (20 if bar else 8), 0xA000 | (sprite + int(bar))])
        next_address = 0x204 + group_bytes * (t + 1)
        if t == len(values) - 1:
            next_address -= 2
        opcodes.extend(
            [
                0x00E0,
                0x6000 | ((14 if bar else 16) + 32 * int(token)),
                0xD011,
                0x1000 | next_address,
            ]
        )
    result = b"".join(value.to_bytes(2, "big") for value in opcodes)
    result += {"train-dot": b"\x80", "heldout-bar": b"\xe0", "nuisance": b"\x80\xe0"}[
        renderer
    ]
    if len(result) > 3584:
        raise ValueError("ROM budget exceeded")
    return result


def validate_rom(renderer, count, rom):
    if renderer not in RENDERERS or not 1 <= count <= 256 or not isinstance(rom, bytes):
        raise ValueError("invalid ROM boundary")
    group = 12 if renderer == "nuisance" else 8
    expected_size = 4 + count * group + (2 if renderer == "nuisance" else 1)
    if len(rom) != expected_size or len(rom) > 3584:
        raise ValueError("ROM shape or budget mismatch")
    tokens = []
    for t in range(count):
        bar = renderer == "heldout-bar" or (renderer == "nuisance" and t % 2 == 1)
        offset = 4 + t * group + (4 if renderer == "nuisance" else 0) + 3
        x = rom[offset] - (14 if bar else 16)
        if x not in (0, 32):
            raise ValueError("noncanonical beacon location")
        tokens.append(x // 32)
    if compile_rom(renderer, tokens) != rom:
        raise ValueError("noncanonical opcode or sprite")


def describe_frame(cells):
    frame = np.asarray(cells)
    if frame.shape != (32, 64) or not np.isin(frame, [0, 1]).all():
        raise ValueError("requires a 64x32 binary frame")
    # Core chooses the lower palette index when occupancy ties.
    background = int(np.count_nonzero(frame) > frame.size // 2)
    points = set(zip(*np.nonzero(frame != background), strict=True))
    if not points:
        raise ValueError("missing beacon")
    first = next(iter(points))
    component, pending = {first}, [first]
    while pending:
        y, x = pending.pop()
        for point in ((y - 1, x), (y + 1, x), (y, x - 1), (y, x + 1)):
            if point in points and point not in component:
                component.add(point)
                pending.append(point)
    if component != points:
        raise ValueError("ambiguous beacon")
    ys, xs = zip(*points, strict=True)
    x0, y0 = min(xs), min(ys)
    if x0 // 32 != max(xs) // 32:
        raise ValueError("beacon crosses midpoint")
    shape = tuple(sorted((int(x - x0), int(y - y0)) for y, x in points))
    return {
        "Token": (sum(int(x) for x in xs) // len(xs)) // 32,
        "Shape": shape,
        "Colour": 1 - background,
        "Origin": (int(x0), int(y0)),
    }


def interpret(renderer, count, rom, on_frame: Callable[[bytes], object] | None = None):
    validate_rom(renderer, count, rom)
    memory = bytearray(4096)
    memory[0x200 : 0x200 + len(rom)] = rom
    registers = [0] * 16
    display = np.zeros((32, 64), dtype=np.uint8)
    pc, index = 0x200, 0

    def step():
        nonlocal pc, index
        if not 0x200 <= pc < 0x200 + len(rom) - 1:
            raise ValueError("PC outside admitted ROM")
        opcode = int.from_bytes(memory[pc : pc + 2], "big")
        pc += 2
        if opcode == 0x00E0:
            display.fill(0)
        elif opcode >> 12 == 6:
            registers[(opcode >> 8) & 15] = opcode & 255
        elif opcode >> 12 == 10:
            index = opcode & 4095
        elif opcode == 0xD011:
            if not 0 <= index < len(memory):
                raise ValueError("sprite outside memory")
            for bit in range(8):
                if memory[index] & (128 >> bit):
                    display[registers[1] % 32, (registers[0] + bit) % 64] ^= 1
        elif opcode >> 12 == 1:
            pc = opcode & 4095
        else:
            raise ValueError("unregistered opcode")

    step()
    step()
    previous = None
    diagnostics = {
        "Comparisons": 0,
        "StructureChanges": 0,
        "PaletteChanges": 0,
        "PlacementChanges": 0,
    }
    tokens = []
    for t in range(count):
        for _ in range(6 if renderer == "nuisance" else 4):
            step()
        frame = display ^ int(renderer == "nuisance" and t % 2 == 1)
        description = describe_frame(frame)
        tokens.append(description["Token"])
        if previous is not None:
            diagnostics["Comparisons"] += 1
            for column, key in (
                ("StructureChanges", "Shape"),
                ("PaletteChanges", "Colour"),
                ("PlacementChanges", "Origin"),
            ):
                diagnostics[column] += int(previous[key] != description[key])
        previous = description
        if on_frame is not None:
            on_frame(frame.tobytes())
    return np.asarray(tokens, dtype=np.uint8), diagnostics


def corpus(renderer, count, length, seed, tag, probability=0.75, start=128, duration=0):
    if not 1 <= count <= 65536:
        raise ValueError("corpus count outside bounded reference")
    stream = Stream(domain(seed, tag))
    rom_hash, frame_hash, token_hash = (hashlib.sha256() for _ in range(3))
    rows = []
    diagnostics = {
        "Comparisons": 0,
        "StructureChanges": 0,
        "PaletteChanges": 0,
        "PlacementChanges": 0,
    }
    for _ in range(count):
        truth = sample(stream, length, probability, start, duration)
        rom = compile_rom(renderer, truth)
        tokens, row_diagnostics = interpret(renderer, length, rom, frame_hash.update)
        if not np.array_equal(tokens, truth):
            raise ValueError("extraction differs from source; no fallback allowed")
        rom_hash.update(rom)
        token_hash.update(tokens.tobytes())
        rows.append(tokens)
        for key in diagnostics:
            diagnostics[key] += row_diagnostics[key]
    return (
        np.asarray(rows),
        {
            "RomSha256": rom_hash.hexdigest().upper(),
            "FrameSha256": frame_hash.hexdigest().upper(),
            "TokenSha256": token_hash.hexdigest().upper(),
            "Sequences": count,
            "Frames": count * length,
        },
        diagnostics,
    )
